File: typescript/lld/booking/hotel/PricingStrategy.ts
import { Room } from './Room';
import { RoomType } from './enums';

// Strategy: nightly rate is pluggable (seasonal, weekend, dynamic).
export interface PricingStrategy {
    perNight(room: Room): number;
}

export class SeasonalPricing implements PricingStrategy {
    constructor(private readonly rates: Record<RoomType, number>) {}
    perNight(room: Room): number { return this.rates[room.type]; }
}

File: java/lld/booking/hotel/PricingStrategy.java
import java.util.Map;

// Strategy: nightly rate is pluggable.
public interface PricingStrategy {
    double perNight(Room room);
}

class SeasonalPricing implements PricingStrategy {
    private final Map<RoomType, Double> rates;
    public SeasonalPricing(Map<RoomType, Double> rates) { this.rates = rates; }
    public double perNight(Room room) { return rates.get(room.type); }
}